crates/matrix-sdk-appservice/src/webserver.rs

```diff
@@ -36,211 +35,157 @@ pub async fn run_server(
-            let user_id = percent_decode_str(&user_id).decode_utf8().map_err(Error::from)?;
-            let request = query_user::IncomingRequest::try_from_http_request(request, &[user_id])
-                .map_err(Error::from)?;
-            return if user_exists(appservice.clone(), request).await {
-                Ok(warp::reply::json(&EmptyObject {}))
+            if user_exists(appservice.clone(), request).await {
+                Ok(Json(EmptyObject {}))
             } else {
-                Err(warp::reject::not_found())
-            };
+                Err(StatusCode::NOT_FOUND)
+            }
+        } else {
+            Ok(Json(EmptyObject {}))
         }
-        Ok(warp::reply::json(&EmptyObject {}))
     }
 
     pub async fn room(
-        room_id: String,
-        appservice: AppService,
-        request: http::Request<Bytes>,
-    ) -> Result<impl Reply, Rejection> {
+        Extension(appservice): Extension<AppService>,
+        MatrixRequest(request): MatrixRequest<query_room_alias::v1::IncomingRequest>,
+    ) -> impl IntoResponse {
         if let Some(room_exists) = appservice.event_handler.rooms.lock().await.as_mut() {
-            let room_id = percent_decode_str(&room_id).decode_utf8().map_err(Error::from)?;
-            let request = query_room::IncomingRequest::try_from_http_request(request, &[room_id])
-                .map_err(Error::from)?;
-            return if room_exists(appservice.clone(), request).await {
-                Ok(warp::reply::json(&EmptyObject {}))
+            if room_exists(appservice.clone(), request).await {
+                Ok(Json(&EmptyObject {}))
             } else {
-                Err(warp::reject::not_found())
-            };
+                Err(StatusCode::NOT_FOUND)
+            }
+        } else {
+            Ok(Json(&EmptyObject {}))
         }
-        Ok(warp::reply::json(&EmptyObject {}))
     }
 
     pub async fn transaction(
-        txn_id: String,
-        appservice: AppService,
-        request: http::Request<Bytes>,
-    ) -> Result<impl Reply, Rejection> {
-        let incoming_transaction: ruma::api::appservice::event::push_events::v1::IncomingRequest =
-            ruma::api::IncomingRequest::try_from_http_request(request, &[txn_id])
-                .map_err(Error::from)?;
-
-        appservice.receive_transaction(incoming_transaction).await?;
-        Ok(warp::reply::json(&EmptyObject {}))
+        appservice: Extension<AppService>,
+        MatrixRequest(request): MatrixRequest<push_events::v1::IncomingRequest>,
+    ) -> impl IntoResponse {
+        match appservice.receive_transaction(request).await {
+            Ok(_) => Ok(Json(&EmptyObject {})),
+            Err(e) => {
+                let status_code = StatusCode::INTERNAL_SERVER_ERROR;
+                Err((
+                    status_code,
+                    Json(ErrorMessage { code: status_code.as_u16(), message: e.to_string() }),
+                ))
+            }
```

Comment:
@johannescpk I'm not clear on what the warp code did when `receive_transaction` failed, can you shed some light on that?